src/ipips/ipip-0476.md

```diff
@@ -0,0 +1,145 @@
+---
+title: "IPIP-0476: Delegated Routing DHT Closest Peers API"
+date: 2025-08-19
+ipip: proposal
+editors:
+  - name: Alex Potsides
+    github: achingbrain
+    affiliation:
+      name: Shipyard
+      url: https://ipshipyard.com
+  - name: Marcin Rataj
+    github: lidel
+    affiliation:
+      name: Shipyard
+      url: https://ipshipyard.com
+relatedIssues:
+  - https://github.com/ipfs/specs/pull/476
+  - https://github.com/ipfs/specs/pull/497
+order: 476
+tags: ['ipips']
+---
+
+## Summary
+
+Add a new HTTP endpoint to the Delegated Routing API that allows clients to find the closest peers to a given peer ID without being full DHT clients.
+
+## Motivation
+
+Browser nodes and other resource-constrained clients need to perform peer discovery operations without the overhead of being full DHT clients. The primary use case is for browser nodes performing random walks to find peers that they can make circuit relay reservations on.
+
+Currently, to find peers close to a particular key in the DHT keyspace, a node must:
+1. Be a full DHT client with all the associated overhead
+2. Maintain connections to many peers
+3. Handle the complexity of the DHT protocol
+
+This is particularly problematic for:
+- Browser nodes that need to find circuit relay servers
+- Light clients that want to populate their routing tables
+- Applications that need to find peers to host provider records
+
+## Detailed design
+
+This IPIP introduces a new "DHT Routing API" section to the Delegated Routing V1 HTTP API specification with the following endpoint:
+
+### `GET /routing/v1/dht/closest/peers/{peer-id}`
+
+#### Path Parameters
+
+- `peer-id`: The target peer ID to find closest peers for, represented as a CIDv1 encoded with `libp2p-key` codec
+
+#### Query Parameters
```

Comment:
Do we have any specific use cases in mind that would require query parameters?

IIUC Go and JS DHT implementations only support returning the `20` closest peers. Since the `FIND_NODE` RPC returns at most `20` peers, it is quite complex to lookup for more than `20` peers.

The query parameters could be used to filter the results and return less than 20 peers, but not to get more peers.

Reply:
@achingbrain @hsanjuan would it be ok to remove `closer-than` and  `count`?

The only implementation (kad-dht) doesn't support these parameters:
- go-libp2p-kad-dht
  - [GetClosestPeers only takes (ctx, key string)](https://github.com/libp2p/go-libp2p-kad-dht/blob/4152c7c0b5a1ca862b4fd6b8034a4f4edd7f2530/lookup.go#L23) 
  - Always returns K (20) closest peers
  - No filtering by closer-than

Imo there is no point in listing them in spec and implementing filtering in userland.